python.fmt.ts

```diff
@@ -42,35 +42,107 @@ export class PythonFormatter extends CodeFormatter {
   indentStr = '    '
   endTypeStr = ''
 
+  // keyword.kwlist
+  pythonKeywords = [
+    'False',
+    'None',
+    'True',
+    'and',
+    'as',
+    'assert',
+    'async',
+    'await',
+    'break',
+    'class',
+    'continue',
+    'def',
+    'del',
+    'elif',
+    'else',
+    'except',
+    'finally',
+    'for',
+    'from',
+    'global',
+    'if',
+    'import',
+    'in',
+    'is',
+    'lambda',
+    'nonlocal',
+    'not',
+    'or',
+    'pass',
+    'raise',
+    'return',
+    'try',
+    'while',
+    'with',
+    'yield'
+  ]
+  pythonTypes: Record<string, IMappedType> = {
+    'number': { name: 'float', default: this.nullStr },
+    'double': { name: 'float', default: this.nullStr },
+    'integer': { name: 'int', default: this.nullStr },
+    'int32': { name: 'int', default: this.nullStr },
+    'int64': { name: 'int', default: this.nullStr },
+    'string': { name: 'str', default: this.nullStr },
+    'password': {name: 'str', default: this.nullStr },
+    'byte': {name: 'bytes', default: this.nullStr },
+    'boolean': {name: 'bool', default: this.nullStr },
+    'void': {name: 'None', default: this.nullStr },
+    'uri': {name: 'str', default: this.nullStr },
+    'datetime': {name: 'datetime.datetime', default: this.nullStr }
+  }
+
+  // Classes requiring custom structure/unstructure
+  reservedKeywordHooks: string[] = [
+    `\n\n# Custom converters for reserved keyword attribute classes`
+  ]
+  forwardRefHooks: string[] = [
+    `\n\n# Custom hooks for type induced forward references\n` +
+    `from typing import ForwardRef  # type: ignore  # noqa:E402\n`
+  ]
+
   // @ts-ignore
   methodsPrologue = (indent: string) => `
 # ${warnEditing}
-from typing import *
-# TODO why do we need "python.looker" for the imports?
-from python.looker.sdk.models import *
-from python.looker.rtl.sdk_base import SDKBase
+import datetime
+from typing import Optional, Sequence
+
+from looker.sdk import models
+from looker.rtl import api_methods
+
 
-class LookerSDK(SDKBase):
+class LookerSDK(api_methods.APIMethods):
 `
   // @ts-ignore
   methodsEpilogue = (indent: string) => ''
   // @ts-ignore
   modelsPrologue = (indent: string) => `
 # ${warnEditing}
+import datetime
+from typing import Optional, Sequence
+
 import attr
-from typing import *
+import cattr
+
+from looker.rtl import model
+from looker.rtl import serialize as sr
 `
   // @ts-ignore
-  modelsEpilogue = (indent: string) => ''
+  modelsEpilogue = (indent: string) => {
+    return this.forwardRefHooks.join('\n') + this.reservedKeywordHooks.join('\n')
+  }
 
   // @ts-ignore
   argGroup(indent: string, args: Arg[]) {
     if ((!args) || args.length === 0) return this.nullStr
     let hash: string[] = []
     for (let arg of args) {
-      hash.push(`'${arg}': ${arg}`)
+      hash.push(`"${arg}": ${arg}`)
     }
-    return `\n${indent}{${hash.join(this.argDelimiter)}}`
+    return `{${hash.join(this.argDelimiter)}}`
```

Comment:
is removal of `indent` because we never expect this to be broken up by line?

Reply:
I forgot to mention that we'd like to add a final step to code generation. We've settled on [black](https://black.readthedocs.io/en/stable/) for code formatting. So, rather than break things up here, I wanted to write it all one one line and then run `black python/looker/sdk/` which will make the code look as pretty as possible :-)

thoughts on how to do this? hook into `IGeneratorSpec.options` somehow?

Reply:
Yeah, I've been thinking about pretty-printing/post-processing (hey, `p`-`p` FTW!) since we started this project. I think we should make a post-process hook as part of the standard code formatter process. We _might_ want to make the post-processor method part of [targetLanguages.ts](targetLanguages.ts), or just make it part of the last process for the language-specific code generators to perform, since they'll have intimate knowledge of the files they write.